packages/nextjs/test/integration/test/utils.js

```diff
@@ -0,0 +1,205 @@
+};
+
+module.exports.isTransactionRequest = request => {
+  return isEnvelopeRequest(request) && extractEnvelopeFromRequest(request).itemHeader.type === 'transaction';
+};
+
+module.exports.waitForAll = actions => {
+  return Promise.all(actions).catch(() => {
+    throw new Error('Failed to await on all requested actions');
+  });
+};
+
+module.exports.sleep = duration => {
+  return new Promise(resolve => setTimeout(() => resolve(), duration));
+};
+
+module.exports.expectEvent = (request, expectedEvent) => {
+  if (!request) throw new Error('Event missing');
+  return assertObjectMatches(extractEventFromRequest(request), expectedEvent);
+};
+
+module.exports.expectSession = (request, expectedItem) => {
+  if (!request) throw new Error('Session missing');
+  const { itemHeader, item } = extractEnvelopeFromRequest(request);
+  return itemHeader.type === 'session' && assertObjectMatches(item, expectedItem);
+};
+
+module.exports.expectTransaction = (request, expectedItem) => {
+  if (!request) throw new Error('Transaction missing');
+  const { itemHeader, item } = extractEnvelopeFromRequest(request);
+  return itemHeader.type === 'transaction' && assertObjectMatches(item, expectedItem);
+};
+
+module.exports.expectRequestCount = (requests, expectedCount, timeout = 100) => {
+  return new Promise((resolve, reject) => {
+    // This is to provide a more human-readable stacktrace instead of truncated one from `setTimeout` call
+    const err = new Error('stacktrace magic tricks');
+    setTimeout(() => {
+      for (const key in expectedCount) {
+        try {
+          strictEqual(requests[key].length, expectedCount[key]);
+        } catch (e) {
+          err.message = e.message;
+          reject(err);
+        }
+      }
+      resolve();
+    }, timeout);
+  });
+};
+
+const extractEventFromRequest = (module.exports.extractEventFromRequest = request => {
+  return JSON.parse(request.postData());
+});
+
+const extractEnvelopeFromRequest = (module.exports.extractEnvelopeFromRequest = request => {
+  return parseEnvelope(request.postData());
+});
+
+const parseEnvelope = body => {
+  const [envelopeHeaderString, itemHeaderString, itemString] = body.split('\n');
+
+  return {
+    envelopeHeader: JSON.parse(envelopeHeaderString),
+    itemHeader: JSON.parse(itemHeaderString),
+    item: JSON.parse(itemString),
+  };
+};
+
+const assertObjectMatches = (actual, expected) => {
+  if (Object.prototype.toString.call(actual) !== Object.prototype.toString.call(expected)) {
+    throw new TypeError(`Types mismatch: ${actual} !== ${expected}`);
+  }
+
+  for (const key in expected) {
+    const expectedValue = expected[key];
+
+    if (Object.prototype.toString.call(expectedValue) === '[object Object]' || Array.isArray(expectedValue)) {
+      assertObjectMatches(actual[key], expectedValue);
+    } else {
+      strictEqual(actual[key], expectedValue);
+    }
+  }
+
+  return true;
+};
+
+// Misc
```

Comment:
I'd separate this misc stuff from the server stuff into two separate files, but this is a minor thing.